boards/frdm-kw41z-k64f: add riotboot

fjmolinas opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

This PR adds riotboot support for kw41z. To do this the following changes had to be done:

- modify tests/cortex_comon_ldscript to also work for kinetis
- add script to prevent bricked kinetis when flashing binary files
- handle rom_offset in kinetis.ld script

I know the linker script related changes should be in another PR but I wanted to get @cladmi feedback on them before splitting them up.

### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

Run ldscript test:

`BOARD=frdm-kw41z APP_VER=$(date +%s) make -C tests/cortexm_common_ldscript/`

Run riotboot tests:
~~BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C examples/hello-world/ riotboot/flash-extended-slot0~~

`BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/xtimer_usleep riotboot/flash-extended-slot0`

~~BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C examples/hello-world/ riotboot/flash-slot1~~

`BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/xtimer_usleep riotboot/flash-slot1 `

### Issues/PRs references

Depends on:
* [x] tests/cortexm_common_ldscript: update code section for kinetis #11588 
* [x] kinetis/fcfield-check*: merge into single file and support binary flash #11589 
* [x] handle rom_offset in kinetis.ld script to split #11630 

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

cladmi commented:
Currently I have issue with the test output:

The board should be whitelisted here:

```
RIOT_CI_BUILD=1 BOARD=frdm-kw41z APP_VER=$(date +%s) make -C tests/cortexm_common_ldscript/
make: Entering directory '/home/harter/work/git/RIOT/tests/cortexm_common_ldscript'
The selected BOARD=frdm-kw41z is not whitelisted: iotlab-a8-m3 iotlab-m3 samr21-xpro arduino-mkr1000 arduino-mkrfox1200 arduino-mkrzero blackpill bluepill feather-m0 opencm904 spark-core


EXPECT ERRORS!


Building application "tests_cortexm_common_ldscript" for "frdm-kw41z" with MCU "kinetis".

   text    data     bss     dec     hex filename
   8928     116    2544   11588    2d44 /home/harter/work/git/RIOT/tests/cortexm_common_ldscript/bin/frdm-kw41z/tests_cortexm_common_ldscript.elf
Test rom offset 1 byte overflow detection: [OK]
Test rom offset substracted from rom length in elffile: [SKIP](Reason: board does not have a ROM_OFFSET configured)
Test compilation with offset 0x1000: [OK]
Test compilation with offset 0x2000: [OK]
Test compilation with half ROM length: [OK]
Test ROM overflow detection (too_big_for_rom): [OK]
Test ROM overflow detection (offset_and_romlen): [OK]
make: Leaving directory '/home/harter/work/git/RIOT/tests/cortexm_common_ldscript'
```

And the flash fails too with `arithmetic expression: expeting primary: " + "`

```
RIOT_CI_BUILD=1 BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C examples/hello-world/ riotboot/flash-extended-slot0
make: Entering directory '/home/harter/work/git/RIOT/examples/hello-world'
"make" -C /home/harter/work/git/RIOT/boards/frdm-kw41z
"make" -C /home/harter/work/git/RIOT/boards/common/kw41z
"make" -C /home/harter/work/git/RIOT/core
"make" -C /home/harter/work/git/RIOT/cpu/kinetis
"make" -C /home/harter/work/git/RIOT/cpu/cortexm_common
"make" -C /home/harter/work/git/RIOT/cpu/cortexm_common/periph
"make" -C /home/harter/work/git/RIOT/cpu/kinetis/periph
"make" -C /home/harter/work/git/RIOT/drivers
"make" -C /home/harter/work/git/RIOT/drivers/periph_common
"make" -C /home/harter/work/git/RIOT/sys
"make" -C /home/harter/work/git/RIOT/sys/checksum
"make" -C /home/harter/work/git/RIOT/sys/newlib_syscalls_default
"make" -C /home/harter/work/git/RIOT/sys/riotboot
"make" -C /home/harter/work/git/RIOT/sys/stdio_uart
compiling /home/harter/work/git/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
/bin/sh: 1: arithmetic expression: expecting primary: " + "
/home/harter/work/git/RIOT/makefiles/boot/riotboot.mk:25: recipe for target '/home/harter/work/git/RIOT/examples/hello-world/bin/frdm-kw41z/hello-world-slot0.elf' failed
make: *** [/home/harter/work/git/RIOT/examples/hello-world/bin/frdm-kw41z/hello-world-slot0.elf] Error 2
make: Leaving directory '/home/harter/work/git/RIOT/examples/hello-world'
```

```
RIOT_CI_BUILD=1 BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C examples/hello-world/ riotboot/flash-slot1 
make: Entering directory '/home/harter/work/git/RIOT/examples/hello-world'
compiling /home/harter/work/git/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
/bin/sh: 1: arithmetic expression: expecting primary: " + "
/home/harter/work/git/RIOT/makefiles/boot/riotboot.mk:25: recipe for target '/home/harter/work/git/RIOT/examples/hello-world/bin/frdm-kw41z/hello-world-slot1.elf' failed
make: *** [/home/harter/work/git/RIOT/examples/hello-world/bin/frdm-kw41z/hello-world-slot1.elf] Error 2
make: Leaving directory '/home/harter/work/git/RIOT/examples/hello-world'
```

And currently, `tests/riotboot` does not work:

```
RIOT_CI_BUILD=1 BOARD=frdm-kw41z make -C tests/riotboot  flash test
make: Entering directory '/home/harter/work/git/RIOT/tests/riotboot'
Building application "tests_riotboot" for "frdm-kw41z" with MCU "kinetis".

"make" -C /home/harter/work/git/RIOT/boards/frdm-kw41z
"make" -C /home/harter/work/git/RIOT/boards/common/kw41z
"make" -C /home/harter/work/git/RIOT/core
"make" -C /home/harter/work/git/RIOT/cpu/kinetis
"make" -C /home/harter/work/git/RIOT/cpu/cortexm_common
"make" -C /home/harter/work/git/RIOT/cpu/cortexm_common/periph
"make" -C /home/harter/work/git/RIOT/cpu/kinetis/periph
"make" -C /home/harter/work/git/RIOT/drivers
"make" -C /home/harter/work/git/RIOT/drivers/periph_common
"make" -C /home/harter/work/git/RIOT/sys
"make" -C /home/harter/work/git/RIOT/sys/checksum
"make" -C /home/harter/work/git/RIOT/sys/newlib_syscalls_default
"make" -C /home/harter/work/git/RIOT/sys/riotboot
"make" -C /home/harter/work/git/RIOT/sys/stdio_uart
compiling /home/harter/work/git/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
creating /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0.riot.bin...
   text    data     bss     dec     hex filename
  12440     500    2608   15548    3cbc /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot.elf
/home/harter/work/git/RIOT/dist/tools/openocd/openocd.sh flash /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0-combined.bin
### Flashing Target ###
arm-none-eabi-objdump: /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0-combined.bin: File format not recognized
arm-none-eabi-objdump: section '.fcfield' mentioned in a -j option, but not found in any input file
Danger of bricking the device during flash!
Flash configuration field of /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0-combined.bin:
arm-none-eabi-objdump: /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0-combined.bin: File format not recognized
arm-none-eabi-objdump: section '.fcfield' mentioned in a -j option, but not found in any input file
Abort flash procedure!
pre-flash checks failed, status=1
/home/harter/work/git/RIOT/tests/riotboot/../../Makefile.include:538: recipe for target 'flash' failed
make: *** [flash] Error 1
make: Leaving directory '/home/harter/work/git/RIOT/tests/riotboot'
```

fjmolinas commented:
@cladmi 

> And the flash fails too with arithmetic expression: expeting primary: " + "

Oh... sorry I forgot that riotboot changed and know you need to `USEMODULE += riotboot_slot` in your application ... I'll open a PR to fix the dependency...

And currently, tests/riotboot does not 

> work:
> RIOT_CI_BUILD=1 BOARD=frdm-kw41z make -C tests/riotboot  flash test

The command is `RIOT_CI_BUILD=1 BOARD=frdm-kw41z make -C tests/riotboot  riotboot/flash test`

cladmi commented on tests/cortexm_common_ldscript/Makefile:
```diff
@@ -90,7 +90,7 @@ test-offset_%: $(BINDIR)/$(APPLICATION)_offset_%.elf
 	$(Q)\
 	TEST_START_ADDR=$$($(PREFIX)readelf --section-headers $^ 2>/dev/null | awk '/.text/{printf "0x%s\n", $$5}'); \
-	EXPECT_START_ADDR=$$(printf "0x%08x" $$(( $(ROM_START_ADDR) + $* ))); \
+	EXPECT_START_ADDR=$$(printf "0x%08x" $$(( $(ROM_START_ADDR) + $(VTOR_OFFSET) + $(FCFIELD_OFFSET) + $* ))); \
```
I find it strange to have here the specific `VTOR_OFFSET` and `FCFIELD_OFFSET`.

Maybe here the issue is to look for `.text` and taking only for the first `PROGBITS` section could work. (output for `tests_cortexm_common_ldscript_offset_0x1000.elf`)

For the `frdm-kw41z`

```
[ 0]                   NULL            00000000 000000 000000 00      0   0  0
[ 1] .vector           PROGBITS        00001000 001000 000400 00   A  0   0  4
```

And for `iotlab-m3`:

```
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        08001000 001000 001ef8 00  AX  0   0  4
```

If they are not needed anymore in this file, they do not need to be set in the Makefile.include at all which is somehow better I think as it is really linkerscript internal specific.

cladmi replied:
I will split this in a separate PR.

fjmolinas replied:
Ok! I was about to dot it, you can tag me to review!

cladmi replied:
https://github.com/RIOT-OS/RIOT/pull/11588

cladmi replied:
With the PR, `VTOR_OFFSET` and `FCFIELD_OFFSET` do not need to be set for this anymore. Currently still for `check-fcfield.sh` but it would be removed.

cladmi replied:
`frdm-kw41z` also needs to be added to the `BOARD_WHITELIST` in this `Makefile`.

cladmi commented:
Somehow I think handling the `.fcfield` is not needed when flashing with an offset, as long as `IMAGE_OFFSET > FCFIELD_END='0x410'`.

https://cache.freescale.com/files/microcontrollers/doc/app_note/AN4507.pdf

For me it seems like this `flash configuration field` is at a fixed address and not relative to the `jump address`. If we can confirm this, it makes things way easier.

And we could just say "We do not want to try flashing a binary if it does not start at 0 or is over `FCFIELD_END`". It somebody really want to handle it in the future, he can verify that the right bits are correct.

Even if the `.fcfield` is not required with an offset, we can still use the same linkerscript anyway, its just "wasting" 40bytes or something.

cladmi commented:
> > And currently, tests/riotboot does not work:
> > RIOT_CI_BUILD=1 BOARD=frdm-kw41z make -C tests/riotboot  flash test
> 
> The command is `RIOT_CI_BUILD=1 BOARD=frdm-kw41z make -C tests/riotboot riotboot/flash test`

It is supposed to work with `flash test` as the makefile is setting `FLASHFILE`:
Thats why `check-field` should know which one to call depending on the file format. I can help on this if you want to do it in make only. Doing it in the script could also work.

```
tests/riotboot/Makefile

...
# Target 'all' will generate the combined file directly.
# It also makes 'flash' and 'flash-only' work without specific command.
FLASHFILE = $(RIOTBOOT_COMBINED_BIN)
```

To make it work in the current version, I also needed to export `RIOTBOOT_HDR_LEN`. But as mentioned inline, I think we do not need to care if `fcfield` is as `0 + fcfieldstart`.

``` diff
diff --git a/sys/riotboot/Makefile.include b/sys/riotboot/Makefile.include
index 4e2d9b6c6..607a1fe1d 100644
--- a/sys/riotboot/Makefile.include
+++ b/sys/riotboot/Makefile.include
@@ -1,7 +1,7 @@
 # Indicate the reserved space for a header, 256B by default
 # Notice that it must be 256B aligned. This is restricted by
 # the Cortex-M0+/3/4/7 architecture
-RIOTBOOT_HDR_LEN ?= 0x100
+export RIOTBOOT_HDR_LEN ?= 0x100

 # By default, slot 0 is found just after RIOTBOOT_LEN. Slot 1 after
 # slot 0. The values might be overridden to add more or less offset
```

And then the test succeed !!!! :+1: 

```
> 2019-05-22 17:31:11,218 - INFO #  main(): This is RIOT! (Version: buildtest)
2019-05-22 17:31:11,227 - INFO # Hello riotboot!
2019-05-22 17:31:11,229 - INFO # You are running RIOT on a(n) frdm-kw41z board.
2019-05-22 17:31:11,230 - INFO # This board features a(n) kinetis MCU.
2019-05-22 17:31:11,232 - INFO # riotboot_test: running from slot 0
2019-05-22 17:31:11,233 - INFO # Image magic_number: 0x544f4952
2019-05-22 17:31:11,237 - INFO # Image Version: 0x5ce56b3a
2019-05-22 17:31:11,239 - INFO # Image start address: 0x00001100
2019-05-22 17:31:11,242 - INFO # Header chksum: 0x431476c1
2019-05-22 17:31:11,243 - INFO #
> 2019-05-22 17:31:11,282 - INFO #  curslotnr
2019-05-22 17:31:11,287 - INFO # Current slot=0
> curslothdr
2019-05-22 17:31:11,348 - INFO #  curslothdr
2019-05-22 17:31:11,350 - INFO # Image magic_number: 0x544f4952
2019-05-22 17:31:11,353 - INFO # Image Version: 0x5ce56b3a
2019-05-22 17:31:11,354 - INFO # Image start address: 0x00001100
2019-05-22 17:31:11,358 - INFO # Header chksum: 0x431476c1
2019-05-22 17:31:11,358 - INFO #
> getslotaddr 0
2019-05-22 17:31:11,418 - INFO #  getslotaddr 0
2019-05-22 17:31:11,419 - INFO # Slot 0 address=0x00001100
> dumpaddrs
2019-05-22 17:31:11,478 - INFO #  dumpaddrs
2019-05-22 17:31:11,483 - INFO # slot 0: metadata: 0x1000 image: 0x00001100
2019-05-22 17:31:11,489 - INFO # slot 1: metadata: 0x40800 image: 0x00000000
```

otherwise I got

```
RIOT_CI_BUILD=1 BOARD=frdm-kw41z make -C tests/riotboot riotboot/flash test
make: Entering directory '/home/harter/work/git/RIOT/tests/riotboot'
compiling /home/harter/work/git/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
creating /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0.riot.bin...
/home/harter/work/git/RIOT/dist/tools/openocd/openocd.sh flash /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0.riot.bin
### Flashing Target ###
/home/harter/work/git/RIOT/cpu/kinetis/dist/check-fcfield-bin.sh: 1: /home/harter/work/git/RIOT/cpu/kinetis/dist/check-fcfield-bin.sh: arithmetic expression: expecting primary: "0x400+"
Danger of bricking the device during flash!
Flash configuration field of /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0.riot.bin:
arm-none-eabi-objdump: /home/harter/work/git/RIOT/tests/riotboot/bin/frdm-kw41z/tests_riotboot-slot0.riot.bin: File format not recognized
Abort flash procedure!
pre-flash checks failed, status=1
```

fjmolinas commented:
> To make it work in the current version, I also needed to export RIOTBOOT_HDR_LEN. But as mentioned inline, I think we do not need to care if fcfield is as 0 + fcfieldstart.

Yes sorry, I had rebased my branch to remove a couple of uneeded commits and removed that. I had realized that but was trying to address all your comments before pushing updates.

> Somehow I think handling the .fcfield is not needed when flashing with an offset, as long as IMAGE_OFFSET > FCFIELD_END='0x410'.

Yes I think you are right,  I'll try to implement!

Thanks for the comments!

cladmi commented on boards/common/frdm/Makefile.include:
```diff
@@ -31,8 +31,11 @@ endif
 
-# Check the flash configuration field before flashing
+ifneq (,$(filter riotboot/%, $(MAKECMDGOALS)))
+export PRE_FLASH_CHECK_SCRIPT = $(RIOTCPU)/$(CPU)/dist/check-fcfield-bin.sh
```
By doing the following,  that use '-bin.sh' if it as .bin file, '-hex.sh' if it is a '.hex' file and the '-elf.sh' by default, you can do `make -C tests/riotboot flash` directly.

```diff
diff --git a/boards/common/frdm/Makefile.include b/boards/common/frdm/Makefile.include
index 81898ef3c..510934ce2 100644
--- a/boards/common/frdm/Makefile.include
+++ b/boards/common/frdm/Makefile.include
@@ -31,11 +31,12 @@ endif
 # Configuration for OpenOCD v0.10.0 and newer
 export OPENOCD_CONFIG ?= $(RIOTBOARD)/common/frdm/dist/openocd-$(CPU_FAMILY).cfg
 
-ifneq (,$(filter riotboot/%, $(MAKECMDGOALS)))
-export PRE_FLASH_CHECK_SCRIPT = $(RIOTCPU)/$(CPU)/dist/check-fcfield-bin.sh
-else
-export PRE_FLASH_CHECK_SCRIPT = $(RIOTCPU)/$(CPU)/dist/check-fcfield-elf.sh
-endif
+# Use the '.bin/.hex' files if a '.bin/.hex' extension is found or the .elf as
+# sane default if not any special case.
+CHECK_FC_FIELD_.bin = $(RIOTCPU)/$(CPU)/dist/check-fcfield-bin.sh
+CHECK_FC_FIELD_.hex = $(RIOTCPU)/$(CPU)/dist/check-fcfield-hex.sh
+CHECK_FC_FIELD_default = $(RIOTCPU)/$(CPU)/dist/check-fcfield-elf.sh
+export PRE_FLASH_CHECK_SCRIPT = $(or $(CHECK_FC_FIELD_$(suffix $(FLASHFILE))),$(CHECK_FC_FIELD_default))
 
 # setup serial terminal
 include $(RIOTMAKE)/tools/serial.inc.mk
```

fjmolinas replied:
I'll merge into a global script.

cladmi commented on cpu/kinetis/dist/check-fcfield-bin.sh:
```diff
@@ -0,0 +1,45 @@
+# This script is supposed to be called from RIOTs
+# unified OpenOCD script (dist/tools/openocd/openocd.sh).
+#
+# Syntax: check-fcfield-hex.sh $HEXFILE
```
```suggestion
# Syntax: check-fcfield-bin.sh $HEXFILE
```

fjmolinas commented on cpu/kinetis/dist/check-fcfield-bin.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
```
Note to self: merge all check-fcfield scripts

cladmi commented:
> > To make it work in the current version, I also needed to export RIOTBOOT_HDR_LEN. But as mentioned inline, I think we do not need to care if fcfield is as 0 + fcfieldstart.
> 
> Yes sorry, I had rebased my branch to remove a couple of uneeded commits and removed that. I had realized that but was trying to address all your comments before pushing updates.

No worry, I play stupid first, execute the command, post the error, and then try to understand :)
At the end, we may not even need to export anything after cleaning up.

It's working already, so it's good :)

fjmolinas commented on cpu/kinetis/dist/check-fcfield-bin.sh:
```diff
@@ -0,0 +1,45 @@
+    exit 2
+fi
+
+if [ -n "${IMAGE_OFFSET}" ]; then
```
Its not really about jumping to invalid offsets or not. What we want to do is verify the content of the fcfield, if its over written or has different content than `fffffffffffffffffffffffffe..ffff` then the flash by get blocked and not be writable. 

I use `IMAGE_OFFSET` because when flashing at an offset the the fcfield is not located at the same position so I need to dissasemble different parts of the bin/hex file.

fjmolinas replied:
No, you can brick it if the values in the fcfield are not the expecte ones. But since they are only read at reset the only ones that are read are in 0x400-0x40f. Si this shouldn't be needed indeed.

cladmi replied:
From the `jumping`  I was refering to the `fcfield` at `jumpaddr + 0x400-0x40f`. Checking if they were considered and if the offseted one could brick it.

But now that I read about these bytes, "prevent reaing rom/prevent writing rom" and that they should be indeed at 0x00000400 there should not be any need to check them for the offset.

Proper handling `0 < IMAGE_OFFSET < 0x500` is not needed for the moment, just providing a "not handled for the moment" would be enough I think.

fjmolinas commented on cpu/kinetis/dist/check-fcfield-bin.sh:
```diff
@@ -0,0 +1,45 @@
+fi
+
+if [ -n "${IMAGE_OFFSET}" ]; then
+    FCFIELD=$((${RIOTBOOT_HDR_LEN}+${VTOR_OFFSET}))
```
So declare another "firmware_offset_inside_the_binary" variable which in this case would be `RIOTBOOT_HDR_LEN`, were would be the adecuate place to define this?

cladmi replied:
No need if we do not check when `IMAGE_OFFSET` is set as you said it https://github.com/RIOT-OS/RIOT/pull/11562#discussion_r287605754

fjmolinas commented:
> Somehow I think handling the .fcfield is not needed when flashing with an offset, as long as IMAGE_OFFSET > FCFIELD_END='0x410'.

> Even if the .fcfield is not required with an offset, we can still use the same linkerscript anyway, its just "wasting" 40bytes or something.

This is true although it would still mean having two linker scripts: one when at offset<0x410 and another when flashing after 0x410. Where the first one would have the .fcfield definition but not the second one .

cladmi commented:
> > Somehow I think handling the .fcfield is not needed when flashing with an offset, as long as IMAGE_OFFSET > FCFIELD_END='0x410'.
> 
> > Even if the .fcfield is not required with an offset, we can still use the same linkerscript anyway, its just "wasting" 40bytes or something.
> 
> This is true although it would still mean having two linker scripts: one when at `offset<0x410` and another when flashing after `0x410` (`cortexm.ld`). Where the first one would have the .fcfield definition but not the second one. Would you be okay with that?

No I would keep using the same linkerscript for simplicity.

I was referring more to checking the `fcfield` in the script.

cladmi commented:
Could you split moving to the common `fcfield.sh` only supporting `elf` and `hex` in a separate PR ?

fjmolinas commented:
> Could you split moving to the common fcfield.sh only supporting elf and hex in a separate PR ?

Will do, do you mind if I push force to take into account all your comments and clean up the commit history?

cladmi commented:
@fjmolinas no problem, I followed the changes and will be able to review the rebased version.

cladmi commented on cpu/kinetis/dist/check-fcfield.sh:
```diff
@@ -0,0 +1,54 @@
+FCFIELD_AWK_REGEX='^ 0400 '
+
+if [ ${FLASHFILE##*.} = bin ]; then
+    OBJDUMP=$(arm-none-eabi-objdump --start-address=${FCFIELD_START} --stop-address=${FCFIELD_END} -bbinary -marm ${FLASHFILE} -s)
```
Better not use `OBJDUMP` as variable name here. In the build system `OBJDUMP` is the name of the tool, and I have a pull request to use `OBJDUMP` from the build system in this file :D (I will do it after this one).

I personally tend to not put commands output into a variable when possible (from this source, in French https://linuxfr.org/news/revue-des-techniques-de-programmation-en-shell). So would more refactor this in a function and use the output with `get_fc_field "${FLASHFILE}" | awk`.
The fact that you need to do `printf %s ${OBJDUMP}` somehow shows the pattern issue.

So at least rename it, and maybe put it in a function

cladmi replied:
Woops, will move to the dedicated pull request :)

cladmi commented on cpu/kinetis/dist/check-fcfield.sh:
```diff
@@ -0,0 +1,54 @@
+if [ ${FLASHFILE##*.} = bin ]; then
+    OBJDUMP=$(arm-none-eabi-objdump --start-address=${FCFIELD_START} --stop-address=${FCFIELD_END} -bbinary -marm ${FLASHFILE} -s)
+elif [ ${FLASHFILE##*.} = elf ]; then
+    OBJDUMP=$(arm-none-eabi-objdump -j.fcfield -s "${ELFFILE}")
```
Here should be using `FLASHFILE`

cladmi commented on cpu/kinetis/dist/check-fcfield.sh:
```diff
@@ -0,0 +1,54 @@
+elif [ ${FLASHFILE##*.} = elf ]; then
+    OBJDUMP=$(arm-none-eabi-objdump -j.fcfield -s "${ELFFILE}")
+elif [ ${FLASHFILE##*.} = hex ]; then
+    OBJDUMP=$(arm-none-eabi-objdump --start-address=${FCFIELD_START} --stop-address=${FCFIELD_END} ${HEXFILE} -s)
```
Here FLASHFILE too.

cladmi commented:
After the linkerscript change, I think you can add all the `kinetis` boards to `BOARD_WHITELIST` in `tests/cortexm_common_ldscript`.

When refactoring, `kinetis.ld` to be more similar to `cortexm.ld`, I think the following changes could be done too:
* I think that somehow all the `LINKFLAGS` with `defsym` in `cpu/kinetis/Makefile.include` can be removed as already defined by `cpu/cortexm_common/Makefile.include` (_ram_base_addr is unused from a git grep search, did not check history yet)
* It could be the right time to put the `_rom_offset/_fw_rom_length` calculation in its own file to not duplicate between both `kinetis.ld` and `cortexm.ld`. 
* The `OUTPUT_FORMAT/OUTPUT_ARCH` may not be needed as they are set in `cortexm_base.ld`.


To put the `riotboot` features, it may be a good idea to check that the same issue as https://github.com/RIOT-OS/RIOT/issues/11274 is not present.

fjmolinas commented:
> To put the riotboot features, it may be a good idea to check that the same issue as #11274 is not present.

It is needed for `frdm-k64f` since its sector/page size is 4k, but not for `frdm-kw41z`. There is another pr that implements flashpage for kinetis #11422. But it depends on including `CPU_FAM` before `Makefile.features` so its still delayed. 

Anyway thats not the point. I can add support for `frdm-k64f` in the PR and that way included `RIOTBOOT_LEN` definition for that case.

> I think that somehow all the LINKFLAGS with defsym in cpu/kinetis/Makefile.include can be removed as already defined by cpu/cortexm_common/Makefile.include (_ram_base_addr is unused from a git grep search, did not check history yet)
> It could be the right time to put the _rom_offset/_fw_rom_length calculation in its own file to not duplicate between both kinetis.ld and cortexm.ld.

Ok will need to look into a it a bit!

fjmolinas commented:
> The OUTPUT_FORMAT/OUTPUT_ARCH may not be needed as they are set in cortexm_base.ld.

I'm not sure about what the precedence between the definition of these and sections or memory declaration must be. I found in some doc (not arm specific ld) `the name of the output section is section. section must meet the constraints of your output format.` Also they are not using the same `OUTPUT_FORMAT`:

kinetis:

```
OUTPUT_FORMAT ("elf32-littlearm", "elf32-bigarm", "elf32-littlearm")
OUTPUT_ARCH(arm)
```

cortex common:
```
OUTPUT_FORMAT("elf32-littlearm", "elf32-littlearm", "elf32-littlearm")
OUTPUT_ARCH(arm)
```

> I think that somehow all the LINKFLAGS with defsym in cpu/kinetis/Makefile.include can be removed as already defined by cpu/cortexm_common/Makefile.include (_ram_base_addr is unused from a git grep search, did not check history yet)

Checked and removed it

> To put the riotboot features, it may be a good idea to check that the same issue as #11274 is not present.

Do you have frdm-k22f? I think this PR should work with frdm-k22f by just adding `FEATURES_PROVIDED+=riotboot`.

> It could be the right time to put the _rom_offset/_fw_rom_length calculation in its own file to not 

Another ld script file?

cladmi commented:
> kinetis:
> 
> ```
> OUTPUT_FORMAT ("elf32-littlearm", "elf32-bigarm", "elf32-littlearm")
> OUTPUT_ARCH(arm)
> ```

My bad, I did not see the difference with `elf32-bigarm`, even in your comment it's only the line length that showed me "double check it is different".



> > It could be the right time to put the _rom_offset/_fw_rom_length calculation in its own file to not
> 
> Another ld script file?

Yes another file. Otherwise we have code duplication.

> Do you have frdm-k22f? I think this PR should work with frdm-k22f by just adding FEATURES_PROVIDED+=riotboot.

Unfortunately I do not have a `frdm-k22f`.

cladmi commented on tests/cortexm_common_ldscript/Makefile:
```diff
@@ -4,18 +4,11 @@ include ../Makefile.tests_common
-BOARD_WHITELIST += iotlab-a8-m3
-BOARD_WHITELIST += iotlab-m3
-BOARD_WHITELIST += samr21-xpro
+BOARD_WHITELIST += frdm-kw41z frdm-k64f frdm-k22f iotlab-a8-m3 iotlab-m3 \
```
Here I prefered the one board per line version for readability.
Also the kinetis` boards should be more with a comment saying "using `cpu/kinetis/kinetis.ld` as they do not use `cortem.ld` directly.

fjmolinas replied:
Ok, I was just going with the structure usually used in tests, but I'll change it.

MrKevinWeiss commented:
Tested on the frdm-k22f board after adding `FEATURES_PROVIDED+=riotboot` to the makefile.features.

Results:
`PORT=/dev/ttyACM1  BOARD=frdm-k22f make flash test -C tests/riotboot`

```
2019-06-03 16:15:48,468 - INFO # main(): This is RIOT! (Version: 2019.07-devel-430-g7f0a17-tstxxx/11562)
2019-06-03 16:15:48,469 - INFO # Hello riotboot!
2019-06-03 16:15:48,473 - INFO # You are running RIOT on a(n) frdm-k22f board.
2019-06-03 16:15:48,476 - INFO # This board features a(n) kinetis MCU.
2019-06-03 16:15:48,479 - INFO # riotboot_test: running from slot 0
2019-06-03 16:15:48,482 - INFO # Image magic_number: 0x544f4952
2019-06-03 16:15:48,484 - INFO # Image Version: 0x5cf52b8d
2019-06-03 16:15:48,487 - INFO # Image start address: 0x00002100
2019-06-03 16:15:48,489 - INFO # Header chksum: 0x648f4724
2019-06-03 16:15:48,490 - INFO # 
> 2019-06-03 16:15:48,576 - INFO #  curslotnr
2019-06-03 16:15:48,577 - INFO # Current slot=0
> curslothdr
2019-06-03 16:15:48,632 - INFO #  curslothdr
2019-06-03 16:15:48,634 - INFO # Image magic_number: 0x544f4952
2019-06-03 16:15:48,637 - INFO # Image Version: 0x5cf52b8d
2019-06-03 16:15:48,640 - INFO # Image start address: 0x00002100
2019-06-03 16:15:48,642 - INFO # Header chksum: 0x648f4724
2019-06-03 16:15:48,642 - INFO # 
> getslotaddr 0
2019-06-03 16:15:48,697 - INFO #  getslotaddr 0
2019-06-03 16:15:48,699 - INFO # Slot 0 address=0x00002100
> dumpaddrs
2019-06-03 16:15:48,753 - INFO #  dumpaddrs
2019-06-03 16:15:48,757 - INFO # slot 0: metadata: 0x2000 image: 0x00002100
2019-06-03 16:15:48,760 - INFO # slot 1: metadata: 0x41000 image: 0x00000000
```

`BOARD=frdm-k22f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C examples/hello-world/ riotboot/flash-slot1`

```
### Flashing Target ###
Value in fcfield is not checked when flashing at $IMAGE_OFFSET >= 0x410
Binfile detected, adding ROM base address: 0x00000000
Flashing with IMAGE_OFFSET: 0x00041000
Open On-Chip Debugger 0.10.0+dev-00399-g09076d1 (2018-04-12-17:12)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "swd". To override use 'transport select <transport>'.
Info : add flash_bank kinetis kx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : CMSIS-DAP: FW Version = 1.0
Info : SWCLK/TCK = 0 SWDIO/TMS = 1 TDI = 0 TDO = 0 nTRST = 0 nRESET = 1
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x2ba01477
Info : MDM: Chip is unsecured. Continuing.
Info : kx.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : Listening on port 35643 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* kx.cpu             cortex_m   little kx.cpu             unknown
Info : MDM: Chip is unsecured. Continuing.
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x00000538 msp: 0x1fff0200
auto erase enabled
Info : Kinetis MK22FN512xxx12 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Info : This device supports Program Longword execution only.
Info : Disabling Kinetis watchdog (initial WDOG_STCTRLH = 0x01d3)
Info : WDOG_STCTRLH = 0x01d2
wrote 10240 bytes from file /home/kevinweiss/WorkingDirectory/RIOT/examples/hello-world/bin/frdm-k22f/hello-world-slot1.riot.bin in 1.203952s (8.306 KiB/s)
verified 9752 bytes in 0.272987s (34.886 KiB/s)
Info : MDM: Chip is unsecured. Continuing.
shutdown command invoked
Done flashing
make: Leaving directory '/home/kevinweiss/WorkingDirectory/RIOT/examples/hello-world'
```

fjmolinas commented:
@MrKevinWeiss Thanks for testing! :)

MrKevinWeiss commented:
That is the consequence of being a hardware hoarder :laughing:

fjmolinas commented:
> My bad, I did not see the difference with elf32-bigarm, even in your comment it's only the line length that showed me "double check it is different".

Looking into the code base I don't think the `-EB -EL` (efl32-bigarm elf32-littlearm) are not being used anyway. So changing `cortexm_base.ld` so `-EB` is `elf32-bigarm` would work. Not of these options are actually used, but if they are to be used it makes more sense that `-EB` be elf32-bigarm.

I read a bunch about this over the weekend and couldn't find any reference to the need or lack off need for declaring ` OUTPUT_FORMAT/OUTPUT_ARCH` at the beginning of a script. Testing everything still works as before, so I think by changing `cortexm_base.ld`  we can remove `OUTPUT_FORMAT/OUTPUT_ARCH`  from kinetis.ld. 

I created a new `cortexm_rom_offset.ld` for the common rom_offset calculation.

Tell me what you think of the changes! If you agree I would probably split the last 5 commits regarding linker script cleanup in another PR

cladmi commented on tests/cortexm_common_ldscript/Makefile:
```diff
@@ -4,11 +4,24 @@ include ../Makefile.tests_common
 # Only tested on these ones for the moment
-BOARD_WHITELIST += frdm-kw41z frdm-k64f frdm-k22f iotlab-a8-m3 iotlab-m3 \
-                   mulle pba-d-01-kw2x samr21-xpro teensy31
+BOARD_WHITELIST += frdm-kw41z
```
Kinetis could still be listed separately for `kinetis.ld` :)

cladmi commented:
I think splitting the linker script related things in a separate PR that makes `tests/cortexm_common_ldscript` pass is good. It shows the dependencies for adding `riotboot`.

I would keep the `OUTPUT_FORMAT/OUTPUT_ARCH` modifications out of this PR/the split for the moment as it is a different type of review to do than the offsets and should trigger other reviewers.

It would leave defining the features and the `RIOTBOOT_LEN` in this PR at the end.

I will update the PR description to show the intermediate steps.

cladmi commented:
I tested on a `pba-d-01-kw2x` and it works on it. (Did not checked if RIOTBOOT_LEN should be updated for alignment).


On the `mulle`, it currently fails due to `RIOTBOOT_LEN` not being big enough I think
```
/opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/bin/ld: /home/harter/work/git/worktree/riot_test/bootloaders/riotboot/bin/mulle/riotboot.elf section `.text' will not fit in region `rom'
/opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/bin/ld: region `rom' overflowed by 1612 bytes
collect2: error: ld returned 1 exit status
/home/harter/work/git/worktree/riot_test/bootloaders/riotboot/../..//Makefile.include:461: recipe for target '/home/harter/work/git/worktree/riot_test/bootloaders/riotboot/bin/mulle/riotboot.elf' failed
```

And the terminal output is quite garbage at the beginning so may be an issue with the double init.

fjmolinas commented:
> I tested on a pba-d-01-kw2x and it works on it. (Did not checked if RIOTBOOT_LEN should be updated for alignment).

I also tested on pba-d-01-kw2x but it doesn't work, there is a problem with clock intialization I still haven't figured out. You probably tested `hello-world` or `tests/riotboot` if you test `tests/xtimer_usleep` you will get a `hardfault`.

cladmi commented:
@fjmolinas Indeed, I get a hardfault with `xtimer_usleep` too.

fjmolinas commented:
@MrKevinWeiss would you mind running this command for frdm-k22f:

`BOARD=frdm-k22f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/xtimer_usleep riotboot/flash`

I realized afterwards that just running tests/riotboot is not always enough, some trouble could still arise, for example pba-d-01-kw2x passes `tests/riotboot` but application using timers don't work in conjunction with riotboot.

MrKevinWeiss commented:
Adding to my todo queue, should be done tomorrow morning.

MrKevinWeiss commented:
I retested on the frdm-k22f board after adding riotboot to features provided and got the following output:
```
BOARD=frdm-k22f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/xtimer_usleep riotboot/flash
make: Entering directory '/home/kevinweiss/WorkingDirectory/RIOT/tests/xtimer_usleep'
compiling /home/kevinweiss/WorkingDirectory/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/boards/frdm-k22f
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/core
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/kinetis
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/kinetis/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/periph_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/auto_init
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/div
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/newlib_syscalls_default
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/stdio_uart
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/xtimer
creating /home/kevinweiss/WorkingDirectory/RIOT/tests/xtimer_usleep/bin/frdm-k22f/tests_xtimer_usleep-slot0.riot.bin...
/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/openocd/openocd.sh flash /home/kevinweiss/WorkingDirectory/RIOT/tests/xtimer_usleep/bin/frdm-k22f/tests_xtimer_usleep-slot0.riot.bin
### Flashing Target ###
Value in fcfield is not checked when flashing at $IMAGE_OFFSET >= 0x410
Binfile detected, adding ROM base address: 0x00000000
Flashing with IMAGE_OFFSET: 0x00002000
Open On-Chip Debugger 0.10.0+dev-00399-g09076d1 (2018-04-12-17:12)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "swd". To override use 'transport select <transport>'.
Info : add flash_bank kinetis kx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : CMSIS-DAP: FW Version = 1.0
Info : SWCLK/TCK = 0 SWDIO/TMS = 1 TDI = 0 TDO = 0 nTRST = 0 nRESET = 1
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x2ba01477
Info : MDM: Chip is unsecured. Continuing.
Info : kx.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : Listening on port 33629 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* kx.cpu             cortex_m   little kx.cpu             running
Info : MDM: Chip is unsecured. Continuing.
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x00000538 msp: 0x1fff0200
auto erase enabled
Info : Kinetis MK22FN512xxx12 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Info : This device supports Program Longword execution only.
Info : Disabling Kinetis watchdog (initial WDOG_STCTRLH = 0x01d3)
Info : WDOG_STCTRLH = 0x01d2
wrote 14336 bytes from file /home/kevinweiss/WorkingDirectory/RIOT/tests/xtimer_usleep/bin/frdm-k22f/tests_xtimer_usleep-slot0.riot.bin in 1.500964s (9.327 KiB/s)
verified 12768 bytes in 0.289993s (42.997 KiB/s)
Info : MDM: Chip is unsecured. Continuing.
shutdown command invoked
Done flashing
Building application "riotboot" for "frdm-k22f" with MCU "kinetis".

"make" -C /home/kevinweiss/WorkingDirectory/RIOT/boards/frdm-k22f
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/core
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/kinetis
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/cortexm_common/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/cpu/kinetis/periph
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/drivers/periph_common
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/checksum
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/newlib_syscalls_default
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/riotboot
"make" -C /home/kevinweiss/WorkingDirectory/RIOT/sys/stdio_uart
   text	   data	    bss	    dec	    hex	filename
   3544	      0	    612	   4156	   103c	/home/kevinweiss/WorkingDirectory/RIOT/bootloaders/riotboot/bin/frdm-k22f/riotboot.elf
/home/kevinweiss/WorkingDirectory/RIOT/dist/tools/openocd/openocd.sh flash /home/kevinweiss/WorkingDirectory/RIOT/bootloaders/riotboot/bin/frdm-k22f/riotboot.elf
### Flashing Target ###
/home/kevinweiss/WorkingDirectory/RIOT/bootloaders/riotboot/bin/frdm-k22f/riotboot.elf is not locked.
Open On-Chip Debugger 0.10.0+dev-00399-g09076d1 (2018-04-12-17:12)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "swd". To override use 'transport select <transport>'.
Info : add flash_bank kinetis kx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : CMSIS-DAP: FW Version = 1.0
Info : SWCLK/TCK = 0 SWDIO/TMS = 1 TDI = 0 TDO = 0 nTRST = 0 nRESET = 1
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x2ba01477
Info : MDM: Chip is unsecured. Continuing.
Info : kx.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : Listening on port 33963 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* kx.cpu             cortex_m   little kx.cpu             running
Info : MDM: Chip is unsecured. Continuing.
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x00000538 msp: 0x1fff0200
auto erase enabled
Info : Kinetis MK22FN512xxx12 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Info : This device supports Program Longword execution only.
Info : Disabling Kinetis watchdog (initial WDOG_STCTRLH = 0x01d3)
Info : WDOG_STCTRLH = 0x01d2
Info : This device supports Program Longword execution only.
wrote 4096 bytes from file /home/kevinweiss/WorkingDirectory/RIOT/bootloaders/riotboot/bin/frdm-k22f/riotboot.elf in 1.008073s (3.968 KiB/s)
verified 3544 bytes in 0.892866s (3.876 KiB/s)
Info : MDM: Chip is unsecured. Continuing.
shutdown command invoked
Done flashing
make: Leaving directory '/home/kevinweiss/WorkingDirectory/RIOT/tests/xtimer_usleep'
```

fjmolinas commented:
@cladmi I found the reason why it wasn't working for pba-d-01-kw2x:

> "ARMv7-M Architecture Reference Manual" B1.5.3 The vector table

> The Vector table must be naturally aligned to a power of two whose alignment value is greater than or equal to (Number of Exceptions supported x 4), with a minimum alignment of 128 bytes.
> 
> "Cortex-M4 Devices Generic User Guide" 4.3.4. Vector Table Offset Register
> 
> 
> The minimum alignment is 32 words, enough for up to 16 interrupts. For more interrupts, adjust the alignment by rounding up to the next power of two. For example, if you require 21 interrupts, the alignment must be on a 64-word boundary because the required table size is 37 words, and the next power of two is 64.

For MKW22D5 the number of INT_VECTORS is (82 + 16), so it need to be `po2(4x98) = 512 (0x200)` aligned.

In this case by default it is located at `RIOTBOOT_LEN + RIOTBOOT_HDR_LEN = 0x1100` which is not aligned to `0x200`. If `RIOTBOOT_LEN` is increased to be `0x1100` it works. But here comes the tricky part... We still need the whole image to start at the beginning of a page. So what should actually be done is increase `RIOTBOOT_HDRL_LEN` to be `0x200` aligned.

This need to be done for all M4, I had struggled a while ago debugging why `riotboot` doesn't work on stm32l4 and its the exact same issue, I'm now able to make it work by changing  `RIOTBOOT_HDRL_LEN`  size to `0x200`.

Anyway I will open separate PR's for this.

cladmi commented:
@fjmolinas Congratulation for finding it.

This details deserve its place too in https://github.com/RIOT-OS/RIOT/tree/master/bootloaders/riotboot on the `RIOTBOOT_HDR_LEN` section

> Please note that `RIOTBOOT_HDR_LEN` depends on the architecture of the
> MCU, since it needs to be aligned to 256B. This is fixed regardless of
> `sizeof(riotboot_hdr_t)`

cladmi commented:
@fjmolinas can you rebase this one on master ? we could get this ones in already.

fjmolinas commented:
@cladmi Rebased and all is green!

cladmi commented:
Thanks, I will re-run the tests once more this afternoon.

cladmi commented on cpu/kinetis/Makefile.include:
```diff
@@ -4,6 +4,13 @@ ifeq (,$(KINETIS_SERIES))
+# If RIOTBOOT_LEN uses an uneven number of flashpages, the remainder of the
+# flash cannot be divided by two slots while staying FLASHPAGE_SIZE aligned.
+ifeq (K, $(KINETIS_SERIES))
+RIOTBOOT_LEN ?= 0x2000
```
Would need two spaces here for indentation
```suggestion
  RIOTBOOT_LEN ?= 0x2000
```

Could you also add that the flashpage size is 4k in the comment ?

You can squash them directly.

cladmi commented:
I could successfully test the `frdm-kw41z`


```
BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/xtimer_usleep riotboot/flash-extended-slot0 term
# Runs the firmware without crash (test currently does not wait for the uart input in master)

VADUZ=1 BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/riotboot riotboot/flash-slot1 
test
# Runs the test successfully

# make slot 1 with a lower version
BOARD=frdm-kw41z FEATURES_REQUIRED+=riotboot APP_VER=1 make -C tests/riotboot riotboot/flash-slot1 term
# Show again the xtimer_usleep output
```

cladmi commented:
However for the `frdm-k64f`, the `xtimer_usleep` currently does not work with the same test


```
BOARD=frdm-k64f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/xtimer_usleep riotboot/flash-extended-slot0 term
# Errors with crash
2019-06-06 14:13:10,140 - INFO # Running test 5 times with 7 distinct sleep times
2019-06-06 14:13:10,144 - INFO # Please hit any key and then ENTER to continue
2019-06-06 14:13:10,154 - INFO #
2019-06-06 14:13:10,165 - INFO # Context before hamain(): This is RIOT! (Version: 2019.07-devel-604-g937276-pr/riot/11562/add_riotboot)
2019-06-06 14:13:10,169 - INFO # Running test 5 times with 7 distinct sleep times
2019-06-06 14:13:10,173 - INFO # Please hit any key and then ENTER to continue
2019-06-06 14:13:10,183 - INFO #
2019-06-06 14:13:10,194 - INFO # Context before hamain(): This is RIOT! (Version: 2019.07-devel-604-g937276-pr/riot/11562/add_riotboot)


BOARD=frdm-k64f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/riotboot riotboot/flash-slot1 test
# Runs the test successfully

# make slot 1 with a lower version
BOARD=frdm-k64f FEATURES_REQUIRED+=riotboot APP_VER=1 make -C tests/riotboot riotboot/flash-slot1 term
# Show again the xtimer_usleep output with the crashes
```

cladmi commented:
For the `frdm-k64f` when setting `RIOTBOOT_HDR_LEN=0x400` I get a few lines of output

```
Type '/exit' to exit.
2019-06-06 14:18:38,196 - INFO # Please hit any key and then ENTER to continue
2019-06-06 14:18:38,198 - INFO # Slept for 10014 us (expected: 10000 us) Offset: 14 us
2019-06-06 14:18:38,200 - INFO # Slept for 50014 us (expected: 50000 us) Offset: 14 us
2019-06-06 14:18:38,204 - INFO # �main(): This is RIOT! (Version: 2019.07-devel-604-g937276-pr/riot/11562/add_riotboot)
2019-06-06 14:18:38,206 - INFO # Running test 5 times with 7 distinct sleep times
2019-06-06 14:18:38,207 - INFO # Please hit any key and then ENTER to continue
2019-06-06 14:18:38,210 - INFO # Slept for 10014 us (expected: 10000 us) Offsepected: 10000 us) Offset: 14 us

# It gets stuck there

# In the output buffer there was still `Slept for 50014 us (expected: 50` but was only shown after resetting the board
```

And if I then reset, I get a normal full output of the `xtimer_usleep` test.

cladmi commented:
We could go for only the `frdm-kw41z` as you originally planned, I Approve that part already.

I did not know there would be more issues with the `k64f` and it does not need to delay more the `frdm-kw41z`.

fjmolinas commented:
@cladmi You are right about frdm-k64f, it's has the same vector table alignment problem. But for me running `tests/xtimer_usleep` is working fine (I pushed a change to redefine `RIOTBOOT_HDR_LEN`). The exact command I'm running is:

`BOARD=frdm-k64f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make clean -C tests/xtimer_usleep/ riotboot/flash`

And then:

`BOARD=frdm-k64f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make -C tests/xtimer_usleep/ test`

```
2019-06-06 15:24:39,136 - INFO # main(): This is RIOT! (Version: 2019.07-devel-616-g6f978-pr_kinetis_riotboot_frdm_kw41z)
2019-06-06 15:24:39,140 - INFO # Running test 5 times with 7 distinct sleep times
2019-06-06 15:24:39,144 - INFO # Please hit any key and then ENTER to continue
2019-06-06 15:24:39,158 - INFO # Slept for 10015 us (expected: 10000 us) Offset: 15 us
a
2019-06-06 15:24:39,212 - INFO # Slept for 50015 us (expected: 50000 us) Offset: 15 us
2019-06-06 15:24:39,226 - INFO # Slept for 10249 us (expected: 10234 us) Offset: 15 us
2019-06-06 15:24:39,287 - INFO # Slept for 56795 us (expected: 56780 us) Offset: 15 us
2019-06-06 15:24:39,303 - INFO # Slept for 12137 us (expected: 12122 us) Offset: 15 us
2019-06-06 15:24:39,406 - INFO # Slept for 98779 us (expected: 98765 us) Offset: 14 us
2019-06-06 15:24:39,485 - INFO # Slept for 75015 us (expected: 75000 us) Offset: 15 us
2019-06-06 15:24:39,499 - INFO # Slept for 10015 us (expected: 10000 us) Offset: 15 us
2019-06-06 15:24:39,553 - INFO # Slept for 50015 us (expected: 50000 us) Offset: 15 us
2019-06-06 15:24:39,567 - INFO # Slept for 10249 us (expected: 10234 us) Offset: 15 us
2019-06-06 15:24:39,628 - INFO # Slept for 56795 us (expected: 56780 us) Offset: 15 us
2019-06-06 15:24:39,645 - INFO # Slept for 12136 us (expected: 12122 us) Offset: 14 us
2019-06-06 15:24:39,747 - INFO # Slept for 98780 us (expected: 98765 us) Offset: 15 us
2019-06-06 15:24:39,826 - INFO # Slept for 75015 us (expected: 75000 us) Offset: 15 us
2019-06-06 15:24:39,840 - INFO # Slept for 10015 us (expected: 10000 us) Offset: 15 us
2019-06-06 15:24:39,894 - INFO # Slept for 50015 us (expected: 50000 us) Offset: 15 us
2019-06-06 15:24:39,908 - INFO # Slept for 10249 us (expected: 10234 us) Offset: 15 us
2019-06-06 15:24:39,969 - INFO # Slept for 56795 us (expected: 56780 us) Offset: 15 us
2019-06-06 15:24:39,985 - INFO # Slept for 12137 us (expected: 12122 us) Offset: 15 us
2019-06-06 15:24:40,088 - INFO # Slept for 98780 us (expected: 98765 us) Offset: 15 us
2019-06-06 15:24:40,167 - INFO # Slept for 75015 us (expected: 75000 us) Offset: 15 us
2019-06-06 15:24:40,181 - INFO # Slept for 10015 us (expected: 10000 us) Offset: 15 us
2019-06-06 15:24:40,235 - INFO # Slept for 50014 us (expected: 50000 us) Offset: 14 us
2019-06-06 15:24:40,249 - INFO # Slept for 10249 us (expected: 10234 us) Offset: 15 us
2019-06-06 15:24:40,310 - INFO # Slept for 56795 us (expected: 56780 us) Offset: 15 us
2019-06-06 15:24:40,326 - INFO # Slept for 12137 us (expected: 12122 us) Offset: 15 us
2019-06-06 15:24:40,429 - INFO # Slept for 98780 us (expected: 98765 us) Offset: 15 us
2019-06-06 15:24:40,508 - INFO # Slept for 75015 us (expected: 75000 us) Offset: 15 us
2019-06-06 15:24:40,522 - INFO # Slept for 10014 us (expected: 10000 us) Offset: 14 us
2019-06-06 15:24:40,576 - INFO # Slept for 50015 us (expected: 50000 us) Offset: 15 us
2019-06-06 15:24:40,590 - INFO # Slept for 10249 us (expected: 10234 us) Offset: 15 us
2019-06-06 15:24:40,651 - INFO # Slept for 56795 us (expected: 56780 us) Offset: 15 us
2019-06-06 15:24:40,667 - INFO # Slept for 12137 us (expected: 12122 us) Offset: 15 us
2019-06-06 15:24:40,770 - INFO # Slept for 98780 us (expected: 98765 us) Offset: 15 us
2019-06-06 15:24:40,849 - INFO # Slept for 75015 us (expected: 75000 us) Offset: 15 us
2019-06-06 15:24:40,851 - INFO # Test ran for 1704919 us

make: Leaving directory '/home/francisco/workspace/RIOT/tests/xtimer_usleep'

```

Everything is as expected. Maybe you need to run `make clean` after changing `RIOTBOOT_HDR_LEN`? I also ran the test with `RIOTBOOT_HDR_LEN ?= 0x400` and it was working the same.

cladmi commented:
@fjmolinas after `make reset` that is done by `make test` it works. The issue I had was without `reset`. Maybe because of the long 
```
Error: timed out while waiting for target halted
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x0000319e psp: 0x1fff0420
Error: error executing cortex_m crc algorithm
verified 528896 bytes in 48.328571s (10.687 KiB/s)
Info : MDM: Chip is unsecured. Continuing.
shutdown command invoked
```
I will see if I can get a clean failure output.

fjmolinas commented:
>  after make reset that is done by make test it works. The issue I had was without reset. 

That is weird, I opened a terminal before flashing and from another terminal executed:

`BOARD=frdm-k64f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make clean -C tests/xtimer_usleep/ riotboot/flash`

The application restes 4 times during flashing, I don't now why but when flashing k64f I always detect an eternal reset after which the flash hangs for a while.

````
Info : Kinetis MK64FN1M0xxx12 detected: 2 flash blocks
Info : 2 PFlash banks: 1024k total
wrote 16384 bytes from file /home/francisco/workspace/RIOT/tests/xtimer_usleep/bin/frdm-k64f/tests_xtimer_usleep-slot0.riot.bin in 0.804033s (19.900 KiB/s)
Info : kx.cpu: external reset detected
Error: timed out while waiting for target halted
target halted due to debug-request, current mode: Thread 
```

But everything succeeds eventually, without any manual `make reset`

cladmi commented:
I re-tried many times on my side and I could not reproduce the issue.
I am thinking it was a situation with the flasher ending up in a weird mode, or me not cleaning everything properly even it should not matter there.

cladmi commented:
I managed to reproduce it again, but not consistently with flashing it two times. with:

```
BOARD=frdm-k64f FEATURES_REQUIRED+=riotboot APP_VER=$(date +%s) make clean -C tests/xtimer_usleep/ riotboot/flash-extended-slot0 term
```

It may be an `openocd` issue too.
```
GNU MCU Eclipse 64-bit Open On-Chip Debugger 0.10.0+dev-00462-gdd1d90111 (2019-01-18-11:37)
```

As it works after `reset` anyway, I would handle it as a tools issue and not of the support.

cladmi commented:
Congratulation for taking care of adding the support from start to end. It has been a long and tough process.

fjmolinas commented:
@cladmi Thanks a lot for the review and very helpfull advice!
